# Barcode QC Software



## Concept:

This script qualifies **_[NEXTFLEXÂ® Unique Dual Index Barcode](https://perkinelmer-appliedgenomics.com/library-preparation-kits/nextflex-ngs-barcodes/nextflex-unique-dual-index-barcodes/)_** purity by aligning NGS data ([fastq reads](https://en.wikipedia.org/wiki/FASTQ_format)) to a barcoded amplicon reference. This aligner uses [Levenshtein](https://en.wikipedia.org/wiki/Levenshtein_distance) automata to account for fuzzy matching of single position mismatched sequences, inserted/deleted sequences, and frame shifted sequences. 

This script emphasizes the use of index filtering. Sequencer induced noise can be filtered away via quality analysis of corresponding index read quality. This script utilizes internal MiSeq index filtering methods to accomplish this task.



## Installation:

Open a Unix terminal.

`git clone -b rhs https://github.com/dfox3/bcqc.git`

Alternatively, you can download the zip from [the current repo](https://github.com/dfox3/bcqc) and unzip the directory.


Files included in bcqc/:
 - **bcqc_remix.py**
 - README.md
 - IONRHSREVCOMP/
    - \*.db references 
 - test_data/
    - ion_order.csv
    - fastqs/
       - \*.fastq.gz



## Requirements:

Unix-based terminal
Python 2.7
 - pip installed with:
    - [biopython](https://biopython.org/)

For more information on how to pip install, visit [this guide](https://packaging.python.org/tutorials/installing-packages/#ensure-you-can-run-pip-from-the-command-line).

 - other imports from [The Python Standard Library](https://docs.python.org/2/library/) (typically pre-installed):
    - [bisect](https://docs.python.org/2/library/bisect.html)
    - [itertools](https://docs.python.org/2/library/itertools.html)
    - [collections](https://docs.python.org/2/library/collections.html)
    - [argparse](https://docs.python.org/2/library/argparse.html)
    - [csv](https://docs.python.org/2/library/csv.html)
    - [gzip](https://docs.python.org/2/library/gzip.html)
    - [datetime](https://docs.python.org/2/library/datetime.html)
    - [sqlite3](https://docs.python.org/2/library/sqlite3.html)
    - [shutil](https://docs.python.org/2/library/shutil.html)

Script may be compatible with Python 3 version, but never tested.

For **index filtering**, the user must have index files (or I files) generated by the MiSeq. To understand how to get a MiSeq to print index files, please view this Illumina's guide: [How to generate FASTQ files for index reads in MiSeq Reporter](https://support.illumina.com/bulletins/2016/06/how-to-create-fastq-files-for-index-reads-in-miseq-reporter.html)



## Use:

Navigate to working bcqc directory.
Execute python script.

```
python bcqc.py --i path/to/in_dir/ --o out_prefix --x

  -h, --help   show this help message and exit
  -i -input    Input directory, contains reads in fastq.gz format. Include
               I1/I2 index files if --x is thrown.
  -o -output   Name of output report. All output reports will contain the
               out_prefix specified. Option available for labeling purposes.
  -a -ampmap   CSV of linked barcode,amplicon numbers
  -t -tag      Name of tag txt file. Default is blah,
  -x           (Recommended) Index filter libraries before quality assessment.
  -d           Distribute mutliple alignments evenly
  -m           Turns on one mismatch per seed length. Levenshtein distance for
               mismatches.
  -l -scanlen  Scan length. Default: 18 (recommended). Ignore if reads are >
               36bp
  -w -windows  Window length. Default: 12 (recommended)
  -s -seed     Seed length. Default: 9 (recommended)

```



### Example for ION TORRENT bcqc:

```
cd bcqc
python2 bcqc_remix.py -i test_data/fastqs/ -o ~/ion_test -a test_data/ion_order.csv -t IONRHSREVCOMP -x -m -s 8 -l 24 -w 40
```



### Output:
Main output:
 - "out_prefix_counts.csv"

The main file output is "out_prefix_counts.csv" (the example output will be in 
"test_counts.csv"). This .csv file contains counts for each barcode within 
individual libraries, and the purity of the barcode within a set of libraries. 

Optional:
 - path/to/in_dir/i/
    - i\*.fastq (not gzipped)
 - "out_prefix_index_filter_report.csv"

Optional output files are generated by activating the -x, or index filtering,
flag. When activated, a folder "i" will be generated in the input directory. It
will contain non-compressed index-filtered fastqs that are used for the barcode
quality assessment. Additionally, there is a filtering report csv name after the 
-o flag.



### Info:

**20210607**

**Dylan Fox**

**dylan.fox@perkinelmer.com**
