# Barcode QC Software



## Concept:

This script qualifies **_[NEXTFLEXÂ® Unique Dual Index Barcode](https://perkinelmer-appliedgenomics.com/library-preparation-kits/nextflex-ngs-barcodes/nextflex-unique-dual-index-barcodes/)_** purity by aligning NGS data ([fastq reads](https://en.wikipedia.org/wiki/FASTQ_format)) to a barcoded amplicon reference. This aligner uses [Levenshtein](https://en.wikipedia.org/wiki/Levenshtein_distance) automata to account for fuzzy matching of single position mismatched sequences, inserted/deleted sequences, and frame shifted sequences. 

This script emphasizes the use of index filtering. Sequencer induced noise can be filtered away via quality analysis of corresponding index read quality. This script utilizes internal MiSeq index filtering methods to accomplish this task.



## Installation:

Open a Unix terminal.

`git clone https://github.com/dfox3/bcqc.git`

Alternatively, you can download the zip from [the current repo](https://github.com/dfox3/bcqc) and unzip the directory.


Files included in bcqc/:
 - nermal.py/c
 - index_filter.py/c
 - **bc_reader.py**
 - README.txt
 - spikedb/
    - \*.db references



## Requirements:

Unix-based terminal
Python 2.7
 - pip installed with:
    - [biopython](https://biopython.org/)

For more information on how to pip install, visit [this guide](https://packaging.python.org/tutorials/installing-packages/#ensure-you-can-run-pip-from-the-command-line) for more information.

 - other imports from [The Python Standard Library](https://docs.python.org/2/library/) (typically pre-installed):
    - [bisect](https://docs.python.org/2/library/bisect.html)
    - [itertools](https://docs.python.org/2/library/itertools.html)
    - [collections](https://docs.python.org/2/library/collections.html)
    - [argparse](https://docs.python.org/2/library/argparse.html)
    - [csv](https://docs.python.org/2/library/csv.html)
    - [gzip](https://docs.python.org/2/library/gzip.html)
    - [datetime](https://docs.python.org/2/library/datetime.html)
    - [sqlite3](https://docs.python.org/2/library/sqlite3.html)
    - [shutil](https://docs.python.org/2/library/shutil.html)

Script may be compatible with Python 3 version, but never tested.

For **index filtering**, the user must have index files (or I files) generated by the MiSeq. To understand how to get a MiSeq to print index files, please view this Illumina's guide: [How to generate FASTQ files for index reads in MiSeq Reporter](https://support.illumina.com/bulletins/2016/06/how-to-create-fastq-files-for-index-reads-in-miseq-reporter.html)



## Use:

Navigate to working bcqc directory.
Execute python script.

```
python bc_reader.py --i path/to/in_dir/ --o out_prefix --x

--h for more help

--i path/to/in_dir/
The input directory must contain only .fastq.gz sequencing files. Include 
I1/I2 index files if --x flag is thrown. 

--o out_prefix
All output reports will contain the out_prefix specified. Option 
available for labelling purposes.

--x (optional, recommended) Index filter libraries before quality assessment.
Requires corresponding index files in --i directory to function properly.

--m (optional) allows 1 mismatch per seed length 
Default: False.

--d (optional) distributes multiple reads to each hit.
Fractions of reads are distributed to tags where there are too many 
mismatches to determine original tag. Default: False.

--l (optional) Seed length. Default: 12 (recommended). 
Minimum: 2. Maximum: 12. Ignore if reads are > 36bp.

--s (optional) Scan length. Default: 32 (recommended). 
Ignore if reads are > 36bp.
```



### Example:

```
cd bcqc
python bc_reader.py --i testFastqs/ --o test --x
```



### Output:
Main output:
 - "out_prefix_counts.csv"

The main file output is "out_prefix_counts.csv" (the example output will be in 
"test_counts.csv"). This .csv file contains counts for each barcode within 
individual libraries, and the purity of the barcode within a set of libraries. 

Optional:
 - path/to/in_dir/i/
    - i\*.fastq (not gzipped)
 - "out_prefix_index_filter_report.csv"

Optional output files are generated by activating the --x, or index filtering,
flag. When activated, a folder "i" will be generated in the input directory. It
will contain non-compressed index-filtered fastqs that are used for the barcode
quality assessment. Additionally, there is a filtering report csv name after the 
--o flag.



### Info:

**20181213**

**Dylan Fox**

**dylan.fox@perkinelmer.com**